RavenDB-18820 Corax - growable buffer for analyzers

MattAszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18820
https://issues.hibernatingrhinos.com/issue/RavenDB-18824

### Additional description

RavenDB-18820

Analyzer requires a buffer before doing work. We don't want to allocate new buffer for each field or to allocate too much so we allocate default buffer. When field is bigger than buffer we just extend buffer.


RavenDB-18824

Changing the approach in dealing with complex objects in Corax. Instead of immediately throwing we'll default to `Indexing: No` for static indexes (unless it was defined otherwise in an index definition)

### Type of change

- Bug fix
- Optimization

### How risky is the change?
- Not relevant

### Backward compatibility


- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works


### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

maciejaszyk committed:
RavenDB-18820 Corax - growable buffer for analyzers.

maciejaszyk committed:
RavenDB-18820 Do not index map-reduce raw entry.

ayende commented on src/Corax/Constants.cs:
```diff
@@ -80,6 +80,11 @@ public enum Operator
 
+        public static class Analyzers
+        {
+            public const int DefaultBufferForAnalyzers = 4 * Size.Kilobyte; //is this enough?
```
That should be plenty big enough for most scenarios, yes.

ayende commented on src/Corax/IndexSearcher/IndexSearcher.cs:
```diff
@@ -139,15 +138,17 @@ internal unsafe ReadOnlySpan<byte> ApplyAnalyzer(ReadOnlySpan<byte> originalTerm
-        Debug.Assert(tokens.Length == 1, $"{nameof(ApplyAnalyzer)} should create only 1 token as a result.");
 
-        return encoded;
+        var buffer = ArrayPool<byte>.Shared.Rent(outputSize); // We can release it only on dispose. todo maciej: write stack of arrays to return.
```
this is now leaked, no?

maciejaszyk replied:
Yes, fixed now.

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -504,6 +508,19 @@ public bool TryDeleteEntry(string key, string term)
 
+        private void UnlikelyGrowBuffer(out Span<byte> buffer, out Span<Token> tokens, int newBufferSize, int newTokenSize)
+        {
+            Debug.Assert(buffer.Length <= newBufferSize);
```
That is valid? It is an out param - it _has_ no length.

maciejaszyk replied:
Good catch

maciejaszyk committed:
RavenDB-18820 Memory leak fix

arekpalinski commented:
Please check failing Debug.Assert

maciejaszyk committed:
RavenDB-18824 Change to Indexing: No when indexing complex field and warn user

maciejaszyk commented:
> Please check failing Debug.Assert

fixed

maciejaszyk committed:
RavenDB-18820 Share Sparrow internals with Corax

arekpalinski committed:
RavenDB-18820 Fixing build

arekpalinski committed:
RavenDB-18824 Adjusting the logic for throwing errors on attempt to index a complex object in Corax:

- for static indexes we throw if `Indexing != No` was defined in a definition
- for auto indexes we throw to immediately let user know about the behavior
- otherwise we default to `Indexing: No`

arekpalinski committed:
RavenDB-18824 Better error message for auto indexes
